src/control/system/membership.go

```diff
@@ -330,22 +330,39 @@ func (m *Membership) HostList(rankSet *RankSet) []string {
 // Members returns slice of references to all system members ordered by rank.
 //
 // Empty rank list implies no filtering/include all and ignore ranks that are
-// not in the membership.
-func (m *Membership) Members(rankSet *RankSet) (members Members) {
+// not in the membership. Optionally filter on desired states.
+func (m *Membership) Members(rankSet *RankSet, desiredStates ...MemberState) (members Members) {
 	m.RLock()
 	defer m.RUnlock()
 
+	mask, _ := MaskFromStates(desiredStates...)
+
 	if rankSet == nil || rankSet.Count() == 0 {
-		var err error
-		members, err = m.db.AllMembers()
-		if err != nil {
-			m.log.Errorf("failed to get all members: %s", err)
-			return nil
+		if mask == AllMemberFilter {
+			// No rank or member filtering required so copy database.
+			var err error
+			members, err = m.db.AllMembers()
+			if err != nil {
+				m.log.Errorf("failed to get all members: %s", err)
+				return nil
+			}
```

Comment:
If this function can now return an error, I'm beginning to feel like we should return the error further up the stack. Are the types of errors returned things we'd actually expect to happen?

Reply:
The function doesn't return an error, I don't really want to change the function signature to return an error as part of this PR.